core/lib/zksync_core/src/metadata_calculator/mod.rs

```diff
@@ -146,6 +151,34 @@ impl MetadataCalculator {
         Ok(GenericAsyncTree::new(db, self.config.mode).await)
     }
 
+    async fn loop_pruning_tree(
+        &self,
+        pool: ConnectionPool<Core>,
+        tree_pruner: MerkleTreePruner<RocksDBWrapper>,
+        pruner_handle: MerkleTreePrunerHandle,
+        stop_receiver: watch::Receiver<bool>,
+    ) -> anyhow::Result<()> {
+        let pruner_thread = tokio::task::spawn_blocking(|| tree_pruner.run());
```

Comment:
This looks too difficult and doesn't look like it'd fit well in the node framework. IMO, you should go for an approach in which all three tasks are publicly exposed and are supposed to be managed by the node framework:

```rust
// Metadata calculator constructor site, e.g. `main.rs` for the EN:
let mut metadata_calculator: MetadataCalculator = ..;
let (tree_pruner, tree_pruning_update_task) = metadata_calculator.enable_pruning(); 
// `tree_pruner` is an async / lazy wrapper for the pruner, similar to `LazyAsyncTreeReader`
task_futures.push(tokio::spawn(tree_pruner.run()));
// `tree_pruning_update_task` is a named wrapper for the logic in this method
task_futures.push(tokio::spawn(tree_pruning_update_task.run(stop_receiver.clone())));
```

Unlike with tree reader, pruner / pruning task are not cloneable, so they would probably need to be passed via oneshot channels (`enable_pruning()` would set the channel senders correspondingly).